src/site/_includes/components/Prerender.js

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const {html} = require('common-tags');
+
+/**
+ * Prerender by adding Speculation Rules JSON
+ * Also adds a GA event so can see how many prerender attempts were made
+ * versus prerender navigations.
+ * @param {string} url The url to prerender
+ * @returns {string}
+ */
+module.exports = (url) => {
+  // Only prerender internal URLs
+  if (
+    !url ||
+    url.toString().startsWith('http://') ||
+    url.toString().startsWith('https://')
+  ) {
+    return '';
+  }
+
+  // Only prerender if browser supports speculationrules
+  // and not in SaveData mode (Chrome should exclude SaveData
+  // browser anyway, but no harm double checking)
+  return html`
+    <script type="module">
```

Comment:
Why create an inline script here? To me it seems simpler to dynamically generate the `<script type="speculationrules">` object itself rather than dynamically generating a script that will then create and insert the `<script type="speculationrules">` element. (Plus it has the benefit of not creating an additional inline script that requires its own CSP entry.)

Also, what if a particular page wants to prerender (or prefetch) more than 1 URL?

Reply:
> Why create an inline script here?

We need to create an inline script anyway to send the GA event further down the code.

I figured by adding this dynamically it would allow us to check some things before adding it, currently this:

```
  // Only prerender if browser supports speculationrules
  // and not in SaveData mode (Chrome should exclude SaveData
  // browser anyway, but no harm double checking)
```

With this check:

```js
      if (
        HTMLScriptElement.supports &&
        HTMLScriptElement.supports('speculationrules') &&
        !navigator?.connection?.saveData
      ) {
```

There's no harm inserting the speculationrules JSON for unsupporting browsers and in theory Chrome should prevent this doing anything in SaveData mode, but don't want to send the GA event in either case, and figured better to do same check for both since I needed an inline script anyway.

> Also, what if a particular page wants to prerender (or prefetch) more than 1 URL?

This can be done in one of two ways:
1. Have multiple URLs in the speculationrules JSON
2. Have multiple speculationrules JSON

2 is easier when different parts of the page might inject this. We could support 1 in the Prerender.js , but it complicates it a little and not sure it will be used that much, so not sure it's worth it. WDYT?

Reply:
The other thing we could do is have a `data-prerender` attribute (or class) on `<a href ...>` elements and have a script run through and pick up all those URLs, dedup them and then add a single speculationrules JSON (and ga events) for them.

That's similar to what we do with `gc-analytics-event` currently.

It would also potentially allow more future prerendering options (e.g. hover or mouse down on a link).

WDYT?

Reply:
> We need to create an inline script anyway to send the GA event further down the code.

There are other ways to do that. Our `analytics.js` module could look for `script[type=speculationrules]` elements and send the events itself. That would be my preference as it keeps all analytics code in one place.

For example, currently, if someone were to move the `partials/analytics.njk` include to the the foot of the page (rather than in the `<head>` like it is now), then the code you have here would silently break. Without any sort of test to ensure this code runs as expected, this is risky.

Reply:
Oh I quite like that. Will change to that.

Reply:
Only thing with that is that it will only work with speculation rules available when analytics.js runs (basically static speculation rules in the source HTML).

Probably OK, but if we start to dynamically inject them later (e.g. on link hover, or mouse down) then need to remember to track analytics separately.